app/src/gecko/java/org/mozilla/focus/web/GeckoViewPrompt.java

```diff
@@ -0,0 +1,882 @@
+        final EditText username;
+        if ((flags & AUTH_FLAG_ONLY_PASSWORD) == 0) {
+            username = new EditText(builder.getContext());
+            username.setHint(R.string.gv_prompt_username_hint);
+            username.setText(options.getString("username"));
+            container.addView(username);
+        } else {
+            username = null;
+        }
+
+        final EditText password = new EditText(builder.getContext());
+        password.setHint(R.string.gv_prompt_password_hint);
+        password.setText(options.getString("password"));
+        password.setInputType(InputType.TYPE_CLASS_TEXT |
+                InputType.TYPE_TEXT_VARIATION_PASSWORD);
+        container.addView(password);
+
+        if (level != AUTH_LEVEL_NONE) {
+            final ImageView secure = new ImageView(builder.getContext());
+            secure.setImageResource(android.R.drawable.ic_lock_lock);
+            container.addView(secure);
+        }
+
+        builder.setNegativeButton(android.R.string.cancel, /* listener */ null)
+                .setPositiveButton(android.R.string.ok,
+                        new DialogInterface.OnClickListener() {
+                            @Override
+                            public void onClick(final DialogInterface dialog, final int which) {
+                                if ((flags & AUTH_FLAG_ONLY_PASSWORD) == 0) {
+                                    callback.confirm(username.getText().toString(),
+                                            password.getText().toString());
+                                } else {
+                                    callback.confirm(password.getText().toString());
+                                }
+                            }
+                        });
+        createStandardDialog(addCheckbox(builder, container, callback), callback).show();
+    }
+
+    private void addChoiceItems(final int type, final ArrayAdapter<GeckoBundle> list,
+                                final GeckoBundle[] items, final String indent) {
+        if (type == CHOICE_TYPE_MENU) {
+            list.addAll(items);
+            return;
+        }
+
+        for (final GeckoBundle item : items) {
+            final GeckoBundle[] children = item.getBundleArray("items");
+            if (indent != null && children == null) {
+                item.putString("label", indent + item.getString("label", ""));
+            }
+            list.add(item);
+
+            if (children != null) {
+                final String newIndent;
+                if (type == CHOICE_TYPE_SINGLE || type == CHOICE_TYPE_MULTIPLE) {
+                    newIndent = (indent != null) ? indent + '\t' : "\t";
+                } else {
+                    newIndent = null;
+                }
+                addChoiceItems(type, list, children, newIndent);
+            }
+        }
+    }
+
+    public void promptForChoice(final GeckoSession session, final String title,
```

Comment:
\[PMD\] INFO: The method 'promptForChoice' has a Standard Cyclomatic Complexity of 33\.